### PR TITLE
Do not serialize types with pointer fields

### PR DESCRIPTION
**Pull request by LingxiaChen**
Implement the Semmle rule: UnmanagedSerialization.ql. 
Use DFS to find if a type contains pointer field directly or indirectly and use a ConcurrentDictionary collection to save all classes has been visited.

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotSerializeTypeWithPointerFields.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                        {
+                            foreach (var pointerField in pointerFields.Keys)
+                            {
+                                compilationAnalysisContext.ReportDiagnostic(
+                                    pointerField.CreateDiagnostic(
+                                        Rule,
+                                        pointerField.Name));
+                            }
+                        });
+
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotSerializeTypeWithPointerFields.cs]: Nit: extra blank line

File: src/Microsoft.NetCore.Analyzers/Core/Security/DoNotSerializeTypeWithPointerFields.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                        {
+                            // Get all the fields can be serialized.
+                            var fields = typeSymbol.GetMembers().OfType<IFieldSymbol>().Where(s => nonSerializedAttribute != null &&
+                                                                                                !s.HasAttribute(nonSerializedAttribute) &&
+                                                                                                !s.IsStatic);
+
+                            foreach (var field in fields)
+                            {
+                                var fieldType = field.Type;
+
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/DoNotSerializeTypeWithPointerFields.cs]: Nit: Another extra blank line

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotSerializeTypeWithPointerFieldsTests.cs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Test.Utilities;
+using Xunit;
+
+namespace Microsoft.NetCore.Analyzers.Security.UnitTests
+{
+    public class DoNotSerializeTypeWithPointerFieldsTests : DiagnosticAnalyzerTestBase
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotSerializeTypeWithPointerFieldsTests.cs]: Would be good to see unit tests for:
- Properties
- Arrays of pointers
- Collections of pointers
- Generic types

Basically all the edge cases tested in DoNotReferSelfInSerializableClassTests.cs.
